tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules2/SetCupStratifiedRule.scala

```diff
@@ -0,0 +1,56 @@
+package at.forsyte.apalache.tla.bmcmt.rules2
+
+import at.forsyte.apalache.tla.bmcmt._
+import at.forsyte.apalache.tla.bmcmt.arena.PtrUtil
+import at.forsyte.apalache.tla.bmcmt.types.CellT
+import at.forsyte.apalache.tla.lir.oper.TlaSetOper
+import at.forsyte.apalache.tla.lir.{OperEx, TlaEx}
+
+/**
+ * Rewrites X \cup Y, that is, a union of two sets (not UNION). In the first encoding, we used a linear number of `in`
+ * queries. However, this happens to be unsound, and we need a quadratic number of queries.
+ *
+ * @author
+ *   Igor Konnov
+ */
+class SetCupStratifiedRule(rewriter: Rewriter) extends StratifiedRule[Unit] {
+
+  def isApplicable(ex: TlaEx, scope: RewriterScope): Boolean = ex match {
+    case OperEx(TlaSetOper.cup, _*) => true
+    case _                          => false
+  }
+
+  def buildArena(ex: TlaEx)(startingScope: RewriterScope): (RewriterScope, ArenaCell, Unit) =
+    ex match {
+      case OperEx(TlaSetOper.cup, leftSet, rightSet) =>
+        // rewrite the set expressions into memory cells
+        val (postLhsScope, leftSetCell) = rewriter.rewrite(leftSet)(startingScope)
+        val (postRhsScope, rightSetCell) = rewriter.rewrite(rightSet)(postLhsScope)
+
+        val leftPtrs = postRhsScope.arena.getHas(leftSetCell)
+        val rightPtrs = postRhsScope.arena.getHas(rightSetCell)
+
+        // Fixed pointers dominate, if no pointer is fixed we take the disjunction of the smt constraints
+        val unionElemPtrs: Seq[ElemPtr] = PtrUtil.mergePtrsByCellMap(leftPtrs ++ rightPtrs)
+
+        // introduce a new set
+        val newType = CellT.fromTypeTag(ex.typeTag)
+        val arenaWithNewCell = postRhsScope.arena.appendCell(newType)
+        val newSetCell = arenaWithNewCell.topCell
+        val arenaWithHasEdges = arenaWithNewCell.appendHas(newSetCell, unionElemPtrs: _*)
+
+        (postRhsScope.copy(arena = arenaWithHasEdges), newSetCell, (leftSetCell, rightSetCell))
```

Comment:
I've actually just recently added exactly this illustration to the ADR:
https://apalache.informal.systems/docs/adr/024adr-arena-pass.html#illustration

Reply:
Aside from that, I anticipate there to be a more significant difference when we get to the rules, which perform less trivial arena operations. Set cup just adds a new cell, and "rewires" has-edges to it, but there's plenty of other rules, which use e.g. cherrypick, where there's a lot more back-and-forth between the arena bits and the SMT bits.

Reply:
I see. Thank you for explaining.

Let's just be on the lookout in case we seem at some point to only be permuting the shape, rather than realizing substantive simplification.

It could be helpful for derisking the design if your next iteration was to take on one of these more substantial rules.